rfd/0152-automatic-database-users-mongodb.md

```diff
@@ -0,0 +1,202 @@
+---
+authors: STeve Huang (xin.huang@goteleport.com)
+state: draft
+---
+
+# RFD 152 - Database Automatic User Provisioning for MongoDB
+
+## Required approvers
+
+Engineering: @r0mant || @smallinsky
+Product: @klizhentas || @xinding33
+Security: @reedloden || @jentfoo
+
+## What
+
+This RFD discusses on how to expand Database Automatic Provisioning feature for
+MongoDB.
+
+## Why
+
+Automatic User Provisioning has been implemented for several SQL databases,
+including PostgreSQL and MySQL, with the basic design described in [RFD
+113](https://github.com/gravitational/teleport/blob/master/rfd/0113-automatic-database-users.md).
+
+Adding support for database user provisioning to MongoDB presents unique
+challenges due to differences in architecture compared to traditional SQL
+databases.
+
+This RFD aims to identify the challenges and provide solutions to address them.
+
+Since the differences in architecture between MongoDB Atlas and self-hosted
+MongoDB are significant, they will be discussed separately in this RFD.
+
+## MongoDB Atlas Details
+
+Automatic User Provisioning will NOT be supported for MongoDB Atlas with the
+reasons discussed below. This RFD should be updated if better solutions are
+found in future iterations.
+
+[Database
+Users](https://www.mongodb.com/docs/atlas/security-add-mongodb-users/) and
+[Custom Database
+Roles](https://www.mongodb.com/docs/atlas/security-add-mongodb-roles/) for
+MongoDB Atlas are managed at a Atlas project level.
+
+As a consequence, database users and roles are NOT modifiable through
+in-database connections. Instead, one can authenticate with Atlas using the
+Atlas SDK and use APIs to manage these database users and roles. Multiple
+deployment jobs will be created to update the MongoDB clusters in this project,
+upon successful APIs calls.
+
+In my personal testing on an Atlas project with a single MongoDB cluster, it
+takes 10~20 seconds for the deployment job to refresh the database user in the
+target MongoDB cluster.
+
+With the current design of Automatic User Provisioning, the database user must
+be updated with new role assignments for each new connection then the roles
+should be revoked once the connection is done. However, waiting for 10+ seconds
+for provisioning the database user each connection will result in a very bad
+user experience (also client may just time out).
+
+## Self-hosted MongoDB Details
+
+The overall flow and logic will follow the previous [RFD
+113](https://github.com/gravitational/teleport/blob/master/rfd/0113-automatic-database-users.md).
+Differences will be outlined in the sections below.
+
+### The admin user connection
+
+The Database Service will connect as an admin user in order to manage database
```

Comment:
Have you considered if a user with high privileges may remove the admin user as a means to maintain their access long term?  Is there anything we can do to make sure that no one can modify our admin user (besides relying on administration on the customer side)?

If not, we should make sure this concern is documented with this feature.

Reply:
This probably applies to all database types right? I don't think it's easy to prevent the modification on our admin user. 

However, when the admin user is misconfigured, a database connection will fail right away as part of auto-user provisioning process, so the user may not be able to connect at all.

If a user with enough high privileges, they can just create other users and authenticate bypassing teleport?

Reply:
I also don't have any recommendations on how we can strengthen this.  I agree documentation is the most important thing to consider here.